// @edsilv/http-status-codes/dist-esmodule/index.js
export var CONTINUE = 100;
export var SWITCHING_PROTOCOLS = 101;
export var PROCESSING = 102;
export var OK = 200;
export var CREATED = 201;
export var ACCEPTED = 202;
export var NON_AUTHORITATIVE_INFORMATION = 203;
export var NO_CONTENT = 204;
export var RESET_CONTENT = 205;
export var PARTIAL_CONTENT = 206;
export var MULTI_STATUS = 207;
export var MULTIPLE_CHOICES = 300;
export var MOVED_PERMANENTLY = 301;
export var MOVED_TEMPORARILY = 302;
export var SEE_OTHER = 303;
export var NOT_MODIFIED = 304;
export var USE_PROXY = 305;
export var TEMPORARY_REDIRECT = 307;
export var BAD_REQUEST = 400;
export var UNAUTHORIZED = 401;
export var PAYMENT_REQUIRED = 402;
export var FORBIDDEN = 403;
export var NOT_FOUND = 404;
export var METHOD_NOT_ALLOWED = 405;
export var NOT_ACCEPTABLE = 406;
export var PROXY_AUTHENTICATION_REQUIRED = 407;
export var REQUEST_TIME_OUT = 408;
export var CONFLICT = 409;
export var GONE = 410;
export var LENGTH_REQUIRED = 411;
export var PRECONDITION_FAILED = 412;
export var REQUEST_ENTITY_TOO_LARGE = 413;
export var REQUEST_URI_TOO_LARGE = 414;
export var UNSUPPORTED_MEDIA_TYPE = 415;
export var REQUESTED_RANGE_NOT_SATISFIABLE = 416;
export var EXPECTATION_FAILED = 417;
export var IM_A_TEAPOT = 418;
export var UNPROCESSABLE_ENTITY = 422;
export var LOCKED = 423;
export var FAILED_DEPENDENCY = 424;
export var UNORDERED_COLLECTION = 425;
export var UPGRADE_REQUIRED = 426;
export var PRECONDITION_REQUIRED = 428;
export var TOO_MANY_REQUESTS = 429;
export var REQUEST_HEADER_FIELDS_TOO_LARGE = 431;
export var INTERNAL_SERVER_ERROR = 500;
export var NOT_IMPLEMENTED = 501;
export var BAD_GATEWAY = 502;
export var SERVICE_UNAVAILABLE = 503;
export var GATEWAY_TIME_OUT = 504;
export var HTTP_VERSION_NOT_SUPPORTED = 505;
export var VARIANT_ALSO_NEGOTIATES = 506;
export var INSUFFICIENT_STORAGE = 507;
export var BANDWIDTH_LIMIT_EXCEEDED = 509;
export var NOT_EXTENDED = 510;
export var NETWORK_AUTHENTICATION_REQUIRED = 511;
