// chartjs-plugin-title-click/chartjs-plugin-titleclick.js
/* chartjs-plugin-titleclick | AlbinoDrought | MIT License | https://github.com/AlbinoDrought/chartjs-plugin-title-click/blob/master/LICENSE */
(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
(function (global){
/**
 * Created by Sean on 4/11/2017.
 */
var Chart = (typeof window !== "undefined" ? window['Chart'] : typeof global !== "undefined" ? global['Chart'] : null);

var handleEvent = function(titleBlock, e) {
    // from core.legend.js
    var me = titleBlock,
        opts = me.options,
        type = e.type === 'mouseup' ? 'click' : e.type;

    if(type === 'mousemove') {
        if(!opts.onHover) return;
    } else if(type === 'click') {
        if(!opts.onClick) return;
    } else {
        return;
    }

    var x = e.x,
        y = e.y;

    var hitBox = {
        left: me.left,
        top: me.top,
        width: me.width,
        height: me.height
    };

    if (x >= hitBox.left && x <= hitBox.left + hitBox.width && y >= hitBox.top && y <= hitBox.top + hitBox.height) {
        // Touching an element
        if (type === 'click') {
            // use e.native for backwards compatibility
            opts.onClick.call(me, e.native, me);
        } else if (type === 'mousemove') {
            // use e.native for backwards compatibility
            opts.onHover.call(me, e.native, me);
        }
    } else if (type === 'mousemove' && opts.onLeave) {
        opts.onLeave.call(me, e.native, me);
    }
};

var titleClickPlugin = {
    afterEvent: function(chartInstance, e) {
        var titleBlock = chartInstance.titleBlock;
        if(titleBlock) {
            handleEvent(titleBlock, e);
        }
    }
};

module.exports = titleClickPlugin;
Chart.pluginService.register(titleClickPlugin);
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}]},{},[1]);
